apollo-router/src/services/trust_dns_connector.rs

```diff
@@ -0,0 +1,64 @@
+use std::future::Future;
+use std::io;
+use std::net::SocketAddr;
+use std::net::ToSocketAddrs;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+use hyper::client::connect::dns::Name;
+use hyper::client::HttpConnector;
+use hyper::service::Service;
+use trust_dns_resolver::TokioAsyncResolver;
+
+/// Wrapper around trust-dns-resolver's
+/// [`TokioAsyncResolver`](https://docs.rs/trust-dns-resolver/0.23.2/trust_dns_resolver/type.TokioAsyncResolver.html)
+///
+/// The resolver runs a background Task which manages dns requests. When a new resolver is created,
+/// the background task is also created, it needs to be spawned on top of an executor before using the client,
+/// or dns requests will block.
+#[derive(Debug, Clone)]
+pub(crate) struct AsyncHyperResolver(TokioAsyncResolver);
+
+impl AsyncHyperResolver {
+    /// constructs a new resolver from default configuration, uses the corresponding method of
+    /// [`TokioAsyncResolver`](https://docs.rs/trust-dns-resolver/0.23.2/trust_dns_resolver/type.TokioAsyncResolver.html#method.new)
+    pub(crate) fn new_from_system_conf() -> Result<Self, io::Error> {
+        let resolver = TokioAsyncResolver::tokio_from_system_conf()?;
+        Ok(Self(resolver))
+    }
+}
+
+impl Service<Name> for AsyncHyperResolver {
+    type Response = std::vec::IntoIter<SocketAddr>;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+    type Error = io::Error;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, name: Name) -> Self::Future {
+        let resolver = self.0.clone();
+
+        let closure = || async move {
```

Comment:
why a closure here?

Reply:
Just porting the code that was already there.

I cleaned it up: fc57d3230bba7ca9f804a80a88307e9c698847d6